## Part 1 - Import and Explore Data

In [1]:
# Connect to Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
#Import Libraries
import pandas as pd
import numpy as np
import datetime
import statsmodels.formula.api as sm

In [3]:
#Import NFL_Game dataset
NFL_Game=pd.read_csv("/content/drive/MyDrive/Sports Performance Analytics Michigan/Course 1 - Foundations of Sports Analytics: Data, Representation, and Models in Sports/Week 4 Content/Week 4 Assignment/nfl_game.csv")
display(NFL_Game)

,game_id,stadium,date,score,weather_temperature,weather_wind_mph,weather_humidity,score_diff,home,win,...,stadium_close,stadium_type,stadium_age,conference,division,season,team_division,team_division_pre2002,team,stadium_neutral
0,1,Orange Bowl,9/2/1966,14,83,6,71,-9,1,0,...,2008.0,outdoor,29,NFC,NFC West,1966,AFC East,AFC East,Miami Dolphins,0
1,1,Orange Bowl,9/2/1966,23,83,6,71,9,0,1,...,2008.0,outdoor,29,AFC,AFC Central,1966,AFC West,AFC West,Oakland Raiders,0
2,2,Rice Stadium,9/3/1966,45,81,7,70,38,1,1,...,NaN,outdoor,16,NFC,NFC West,1966,NaN,AFC Central,Houston Oilers,0
3,2,Rice Stadium,9/3/1966,7,81,7,70,-38,0,0,...,NaN,outdoor,16,NFC,NFC West,1966,AFC West,AFC West,Denver Broncos,0
4,3,Balboa Stadium,9/4/1966,27,70,7,82,20,1,1,...,NaN,outdoor,52,NFC,NFC West,1966,AFC West,AFC West,San Diego Chargers,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
24309,12155,MetLife Stadium,12/29/2019,17,42,5,69,-17,1,0,...,NaN,outdoor,9,NFC,NFC East,2019,NFC East,NFC East,New York Giants,0
24310,12156,CenturyLink Field,12/29/2019,26,51,3,70,5,0,1,...,NaN,outdoor,17,AFC,AFC West,2019,NFC West,NFC West,San Francisco 49ers,0
24311,12156,CenturyLink Field,12/29/2019,21,51,3,70,-5,1,0,...,NaN,outdoor,17,AFC,AFC West,2019,NFC West,AFC West,Seattle Seahawks,0
24312,12157,Raymond James Stadium,12/29/2019,28,78,12,77,6,0,1,...,2016.0,outdoor,21,NFC,AFC East,2019,NFC South,NFC West,Atlanta Falcons,0


In [4]:
# Calculate summary statistics for the date variable
NFL_Game['date'].describe()

,date
count,24314
unique,2065
top,12/29/2019
freq,32


From our summary statistics we can see that there are 24,314 dates in the column, there are 2065 uniques dates in the column, the top date was 12/29/2019 and it showed up 32 times in the column.

In [5]:
# Calculate summary statistics for score variable based on if the game was home or away
NFL_Game.groupby('home')['score'].describe()

,count,mean,std,min,25%,50%,75%,max
home,,,,,,,,
0,12157.0,19.643004,10.166614,0.0,13.0,20.0,27.0,62.0
1,12157.0,22.254997,10.533005,0.0,14.0,21.0,29.0,72.0


From these statistics we can see that there are the same number of home and away games.  The mean score for a home team was 22.25 and the mean score for an away team was 19.643.  The home team scores range from 0 to 72 with a median at 21 and the away team scores range from 0 to 62 with a median at 20.

Let's check a few correlation coefficients before getting into the regression.

In [6]:
# Correlation coefficient between win and home
NFL_Game['win'].corr(NFL_Game['home'])

np.float64(0.14790211753177582)

In [7]:
# Correlation coefficient between score_diff and home
NFL_Game['score_diff'].corr(NFL_Game['home'])

np.float64(0.1725059685987271)

In [8]:
# Correlation coefficient between score and weather_temperature
NFL_Game['score'].corr(NFL_Game['weather_temperature'])

np.float64(0.03361690080893199)

In [9]:
# Correlation coefficient between score and weather_humidity
NFL_Game['score'].corr(NFL_Game['weather_humidity'])

np.float64(-0.03278832087607435)

In [10]:
# Correlation coefficient between score and weather_wind_mph
NFL_Game['score'].corr(NFL_Game['weather_wind_mph'])

np.float64(-0.07895602955806046)

## Part 2 - Regression Analysis 1 - Test of Home Game Advantage

In this regression analysis, you will try to determine if playing at home gives teams any advantage in their performance.

Run the following regressions where the variable “score_diff” is the dependent variable.

In [11]:
# Regression of score_diff on home
Reg1_1 = sm.ols(formula = 'score_diff ~ home', data= NFL_Game).fit()
print(Reg1_1.summary())

                            OLS Regression Results                            
Dep. Variable:             score_diff   R-squared:                       0.030
Model:                            OLS   Adj. R-squared:                  0.030
Method:                 Least Squares   F-statistic:                     745.7
Date:                Fri, 31 Jul 2026   Prob (F-statistic):          9.64e-162
Time:                        17:33:57   Log-Likelihood:            -1.0020e+05
No. Observations:               24314   AIC:                         2.004e+05
Df Residuals:                   24312   BIC:                         2.004e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -2.6120      0.135    -19.309      0.0

Our regression tells us that if a team is home the score diff will increase by 5.22.  We also have a high t-value and p-value less than 0.05 indicating that this is statistically significant.  The R-Squared however is only 0.030 indicating that our regression only captures about 3 percent of variation in the data.  We can definitely get a better model which we will try to do with the second regression.

In [ ]:
# Regression of score_diff on home, stadium_capacity, stadium_neutral, and the interaction between home and stadium neutral
Reg1_2 = sm.ols(formula = 'score_diff ~ home+stadium_capacity+stadium_neutral+home*stadium_neutral', data= NFL_Game).fit()
print(Reg1_2.summary())

                            OLS Regression Results                            
Dep. Variable:             score_diff   R-squared:                       0.031
Model:                            OLS   Adj. R-squared:                  0.030
Method:                 Least Squares   F-statistic:                     191.7
Date:                Tue, 28 Jul 2026   Prob (F-statistic):          3.99e-162
Time:                        21:52:27   Log-Likelihood:            -1.0019e+05
No. Observations:               24314   AIC:                         2.004e+05
Df Residuals:                   24309   BIC:                         2.004e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept               -2.6386 

Our regression tells us

*   If a team is home the score diff will increase by 5.28.
*   If the stadium capacity increases by one the score diff will increase by 0.0000000052
*   If a game is at a neutral site the score diff will increase by 7.25.
*   If the team is a home team and the game is played at a neutral site the score diff will be 5.2765-14.5038 = -9.2273 meaning it will decrease by 9.2273


We also have high t-value and p-value less than 0.05 for all variables except stadium capacity indicating that all the interactions are statistically significant except for stadium capacity. The R-Squared only increases 0.031 indicating that our regression only captures about 3.1 percent of variation in the data. This is better than our first model by a slim margin but maybe our last model of this type can get us a better interpretation of the data.

In [12]:
# Regression of score_diff on home, stadium_capacity, stadium_neutral, the interaction between home and stadium neutral and the interaction between team and opponent.
Reg1_3 = sm.ols(formula = 'score_diff ~ home+stadium_capacity+stadium_neutral+home*stadium_neutral+team*opponent', data= NFL_Game).fit()
print(Reg1_3.summary())

                            OLS Regression Results                            
Dep. Variable:             score_diff   R-squared:                       0.128
Model:                            OLS   Adj. R-squared:                  0.073
Method:                 Least Squares   F-statistic:                     2.321
Date:                Fri, 31 Jul 2026   Prob (F-statistic):          1.34e-137
Time:                        18:13:10   Log-Likelihood:                -98906.
No. Observations:               24314   AIC:                         2.007e+05
Df Residuals:                   22866   BIC:                         2.124e+05
Df Model:                        1447                                         
Covariance Type:            nonrobust                                         
                                                                    coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------

This regression shows us the statistics for the variables explored in the last model but adds the interaction between every team and opponent which gives us a lot more variables.  Looking at the R-squared we can see it increased from the 0.031 we got in our last model to 0.128 meaning this model captures 12.8 percent of variation.  This model is about 4 times better than our last one.  Now we will explore the impact of outside factors on score.

## Part 3 - Regression Analysis 2 -- Impact of Outside Factors on Scores

In this regression analysis, you will investigate if the final score earned by each team is affected by outside factors such as the size and condition of the stadium as well as weather conditions at the stadium.

Run the following regressions where the variable “score” is the dependent variable.

In [13]:
# Regression of score on season and home
Reg2_1 = sm.ols(formula = 'score ~ season+home', data= NFL_Game).fit()
print(Reg2_1.summary())

                            OLS Regression Results                            
Dep. Variable:                  score   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.022
Method:                 Least Squares   F-statistic:                     273.3
Date:                Fri, 31 Jul 2026   Prob (F-statistic):          4.37e-118
Time:                        18:35:19   Log-Likelihood:                -91246.
No. Observations:               24314   AIC:                         1.825e+05
Df Residuals:                   24311   BIC:                         1.825e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -88.2969      8.615    -10.249      0.0

Our regression tells us that if a team is home their score will increase by 2.61 points. Also, as the season increases by one year the score will increase by 0.054 points.  We have a high t-value and p-value less than 0.05 indicating that these relationships are statistically significant. The R-Squared however is only 0.022 indicating that our regression only captures about 2.2 percent of variation in the data. We can definitely get a better model which we will try to do with the second regression.

In [14]:
# Regression of score on season, home, weather_temperature, weather_wind_mph, and weather_humidity
Reg2_2 = sm.ols(formula = 'score ~ season+home+weather_temperature+weather_wind_mph+weather_humidity', data= NFL_Game).fit()
print(Reg2_2.summary())

                            OLS Regression Results                            
Dep. Variable:                  score   R-squared:                       0.026
Model:                            OLS   Adj. R-squared:                  0.026
Method:                 Least Squares   F-statistic:                     129.9
Date:                Fri, 31 Jul 2026   Prob (F-statistic):          2.60e-136
Time:                        18:42:46   Log-Likelihood:                -91195.
No. Observations:               24314   AIC:                         1.824e+05
Df Residuals:                   24308   BIC:                         1.825e+05
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept             -58.1972    

Our regression tells us that

*   If a team is home their score will increase by 2.61 points.
*   As the season increases by one year the score will increase by 0.04 points.
*   As the temperature increases by one degree score will decrease by 0.0006 points.
*   As the wind speed increases by one mph score will decrease by 0.1188 points.
*   As the humidity increases by one percent score will decrease by 0.015 points.

We have a high t-value and p-value less than 0.05 for all variables besides weather_temperature indicating that these relationships are all statistically significant except for temperature. The R-Squared however increases slightly to 0.026 indicating that our regression only captures about 2.6 percent of variation in the data. We will try to get a better model in our third regression.

In [15]:
# Regression of score on season, home, weather_temperature, weather_wind_mph, weather_humidity, stadium_capacity, stadium_age, stadium_type, stadium_neutral, and the interaction between home and stadium_neutral
Reg2_3 = sm.ols(formula = 'score ~ season+home+weather_temperature+weather_wind_mph+weather_humidity+stadium_capacity+stadium_age+stadium_type+stadium_neutral+home*stadium_neutral', data= NFL_Game).fit()
print(Reg2_3.summary())

                            OLS Regression Results                            
Dep. Variable:                  score   R-squared:                       0.027
Model:                            OLS   Adj. R-squared:                  0.026
Method:                 Least Squares   F-statistic:                     60.80
Date:                Fri, 31 Jul 2026   Prob (F-statistic):          2.09e-134
Time:                        19:00:41   Log-Likelihood:                -91186.
No. Observations:               24314   AIC:                         1.824e+05
Df Residuals:                   24302   BIC:                         1.825e+05
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept         

Our regression tells us that

*   If a team is home their score will increase by 2.64 points.
*   As the season increases by one year the score will increase by 0.039 points.
*   As the temperature increases by one degree score will decrease by 0.0009 points.
*   As the wind speed increases by one mph score will decrease by 0.1329 points.
*   As the humidity increases by one percent score will decrease by 0.0144 points.
*   If a team plays in an outdoor stadium their score will increase by 0.326 points.
*   If a team plays in a retractable stadium their score will increase by 0.597 points.
*   If the stadium increases by one person, the team's score will decrease by 0.0000124 points.
*   As the stadium age increases by one year, the team's score will increase by 0.0053 points.
*   If the team plays at a neutral site, their score will increase by 3.85 points
*   If the home team plays at a neutral site, their score decrease by (2.6382-7.2545 = -4.6163) 4.616 points.

We have a high t-value and p-value less than 0.05 for season, home, weather_wind_mph, weather_humidity, stadium_neutral, and home*stadium_neutral indicating that these relationships are all statistically significant where all other variables present are not statistically significant. The R-Squared again increases slightly from 0.26 to 0.27 indicating that our regression only captures about 2.7 percent of variation in the data. We will try one last time to get a better model in our final regression.

In [16]:
# Regression of score on season, home, weather_temperature, weather_wind_mph, weather_humidity, stadium_capacity, stadium_age, stadium_type, stadium_neutral, team, opponent and the interaction between home and stadium_neutral
Reg2_4 = sm.ols(formula = 'score ~ season+home+weather_temperature+weather_wind_mph+weather_humidity+stadium_capacity+stadium_age+stadium_type+stadium_neutral+home*stadium_neutral+team+opponent', data= NFL_Game).fit()
print(Reg2_4.summary())

                            OLS Regression Results                            
Dep. Variable:                  score   R-squared:                       0.052
Model:                            OLS   Adj. R-squared:                  0.048
Method:                 Least Squares   F-statistic:                     14.52
Date:                Fri, 31 Jul 2026   Prob (F-statistic):          9.04e-211
Time:                        19:37:22   Log-Likelihood:                -90870.
No. Observations:               24314   AIC:                         1.819e+05
Df Residuals:                   24222   BIC:                         1.827e+05
Df Model:                          91                                         
Covariance Type:            nonrobust                                         
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
Intercep

This regression shows us the statistics for the variables explored in the last model but adds the team and opponent which allows us to look at an increase or decrease in score based on which team is playing and who their opponent is. Looking at the R-squared we can see it increased from the 0.027 we got in our last model to 0.052 meaning this model captures 5.2 percent of variation. This model is about 2 times better than our last one.

In [20]:
NFL_Game['stadium_type'].value_counts()

,count
stadium_type,
outdoor,19510
indoor,3892
retractable,912
